### ISSUE TITLE
Update Python minor version from 3.9 to something newer

### ISSUE THREAD

**Issue by jamesobutler** [Slicer/Slicer#7060]
## Is your feature request related to a problem? Please describe.

As I learned while working on #7046, the latest versions of python packages such as `numpy` and `scipy` now have a python minimum of 3.9 which is the version currently used by Slicer (3.9.10). Per [NEP29 drop schedule](https://numpy.org/neps/nep-0029-deprecation_policy.html#drop-schedule), Python 3.9 support will begin to be dropped by major python packages beginning April 5th, 2024. Writing up this issue is to stay ahead of the deprecation curve so that Slicer does not get into a situation of not being able to use the latest versions of major python packages.

## To answer the question: Is the newer version of Python supported by the open source community?
When Slicer was updated from 3.6.7 to 3.9.10 in https://github.com/Slicer/Slicer/commit/34e48e8aef5dad19ec8a955d6f48a1940846e3f3:
It was January 25th, 2022 and the "[python readiness](https://pyreadiness.org/)" for Python 3.9 was that [64.7%](https://github.com/di/pyreadiness/blob/50d28f9bcf62edf5f42e237082f1a3c93f8be3e1/docs/3.9/index.html#L69) of the most popular python packages had explicit support for 3.9 as stated on pypi.org.

As of June 28th, 2023 (today), the "python readiness" for Python 3.10 is that [72.8%](https://github.com/di/pyreadiness/blob/a7f95961c5615d5eb336ad646fe3d1aae8cf8c55/docs/3.10/index.html#L95) of the most popular python packages have explicit support for 3.10 as stated on pypi.org.

As of June 28th, 2023 (today), the "python readiness" for Python 3.11 is that [51.7%](https://github.com/di/pyreadiness/blob/a7f95961c5615d5eb336ad646fe3d1aae8cf8c55/docs/3.11/index.html#L95) of the most popular python packages have explicit support for 3.11 as stated on pypi.org.

## Describe the solution you'd like

- Update Slicer's Python from 3.9 to 3.10. 

Based on Slicer's previous update to Python 3.9 at 64.7% readiness and Python 3.10 at 72.8% readiness as of June 28th, 2023 (today), this means Slicer can consider Python 3.10 ready to update to pending Python 3.10 support at https://github.com/python-cmake-buildsystem/python-cmake-buildsystem. https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/pull/281

## Describe alternatives you've considered

- Update Slicer's Python from 3.9 to 3.11. This should be considered especially if the python-readiness for Python 3.11 (https://pyreadiness.org/3.11/) increases when a developer begins to execute on this update.

**Comment by jamesobutler** [comment]
Scikit-image development has dropped Python 3.9 (https://github.com/scikit-image/scikit-image/commit/4841d5358e8cb4015a692411afa9e18de4e5e001)

**Comment by lassoan** [comment]
We should have a policy for when to update Python version and packages. Similar to this: https://scientific-python.org/specs/spec-0000/

For example, Python major.minor version should be the latest that is out for at least 1 year. For Python packages, we could use major.minor that has been our for 6 months. Always update to the latest patch version in every quarter.

**Comment by jamesobutler** [comment]
It appears that Numpy development has dropped support for Python 3.9 https://github.com/numpy/numpy/commit/2452ff4f7d2cf7ede3b50945fce2d4fe194f7a0e.

**Comment by hubutui** [comment]
meanwhile, `imp` is deprecated since version 3.4 according to [this doc](https://docs.python.org/3.10/library/imp.html). Slicer still use it, eg.: https://github.com/Slicer/Slicer/blob/145cf89213a26514c7bce107df706e5f6396334b/Base/Python/slicer/__init__.py#L131

**Comment by jamesobutler** [comment]
Numpy 2.1.0 (released 2024-08-24) dropped support for Python 3.9. https://github.com/numpy/numpy/releases/tag/v2.1.0
Scipy 1.14.0 (released 2024-06-24) dropped support for Python 3.9. https://github.com/scipy/scipy/releases/tag/v1.14.0

**Comment by guillefix** [comment]
Is there any progress on this?

**Comment by lassoan** [comment]
Yes, there is. We'll release Slicer‐5.8 with Python-3.9 in a few weeks and update Python in Slicer very soon after that.

**Comment by jamesobutler** [comment]
I have updated the "Solution I'd Like" in the original post to make it clear that going to Python 3.10 is no longer desired as the scientific python community has already begun dropping support for Python 3.10 in their latest development versions.

Scikit-image development has dropped Python 3.10 (https://github.com/scikit-image/scikit-image/commit/a7800a458069c146ee5c7d1306f3fd47648ecdcf)
Numpy development has dropped Python 3.10 (https://github.com/numpy/numpy/commit/688669f81c13fc98671db97a464a57604d4e1a0b)

**Comment by lassoan** [comment]
It does not matter for us what Python versions are supported by low-level libraries such numpy or scikit-image, because we don't directly rely on any of their new features. Instead, we need to make decisions based on what Python versions are supported by higher-level packages that important in our field, such as PyTorch, MONAI, nnUnet, ITKPython. We should use Python versions that they all already support.

It would be great if you could identify a full list of key Python packages and then write a script that queries PyPI for available releases. We could then use that script to easily identify the range of Python versions that Slicer could work with.

**Comment by jamesobutler** [comment]
For my Slicer custom application, using near latest versions of Python packages is required to stay current with latest security patch releases. If a package doesn’t support the latest security release then I would work to update the various packages to support this new versions.

**Comment by lassoan** [comment]
> For my Slicer custom application, using near latest versions of Python packages is required to stay current with latest security patch releases

What is the reason for this? Is Slicer used in an untrusted environment? What is the risk that is mitigated by using Python packages with latest security patches?

> If a package doesn’t support the latest security release then I would work to update the various packages to support this new versions.

Do you mean that your experience is that package maintainers are reluctant to release security patches before the [CPython versions reaches `end-of-life`](https://devguide.python.org/versions/)? At what point does that happen - when a CPython version reaches `bugfix` or `security` phase?

**Comment by jamesobutler** [comment]
For our group, our digital products team dictates that we cannot release if Snyk identifies a risk with a high severity or higher. Reviewing https://security.snyk.io/package/pip/torch there was a high severity issue affecting torch versions 2.1.2 and earlier. Torch maintainers did not release a 2.1.3 with the fix as the security fix only came starting with 2.2.0. 

Python package maintainers release security patches that sometime come only in a new minor release. That minor release can correspond with a drop in Python version or other compatibility issues such as drop in platform. Maintainers may only support patch releases for their most recent major.minor release. The Scientific Python follows [SPEC0](https://scientific-python.org/specs/spec-0000/) which drops Python versions sooner than the Python org states CPython as being end-of-life. To avoid the issue of requiring a Python version update to resolve a Python package security issue, staying up to latest maintained versions of Python packages helps resolve the security issue quickly for our customers.

> What is the risk that is mitigated by using Python packages with latest security patches?

Our corporate reputation is at risk for using components with documented high severity security releases that our customers and their own enterprise IT departments are reviewing.

**Comment by lassoan** [comment]
Thanks for the SPEC0 link, it is useful. We could incorporate SPEC0's Python version deprecation timeline (not use Python for more than 3 years after initial release) into our Python update guidelines. We still need to look at our main dependencies to decide on how recent Python version we can jump to when it's time to update.

> at risk for using components with documented high severity security releases

Medical image computing libraries are expected to be used only on trusted data (data that is not carefully crafted by malicious actors to make software misbehave). Even [numpy is advised to be used only on trusted data](https://numpy.org/doc/stable/reference/security.html#advice-for-using-numpy-on-untrusted-data). So, while keeping Python and packages up-to-date is useful, it is highly unlikely to actually affect the security of open-source medical image computing software on untrusted data or in an untrusted environment.

**Comment by jamesobutler** [comment]
Yes I think we should follow SPEC0 to help guide the python version to use in Slicer. If we identify compatibility issues with frequently used python packages in Slicer when going to update the python version then we should aim to help update those packages to support the newer python versions. At this point in time, if a package doesn't support Python 3.11 or Python 3.12 which were released 2 years and 1 year ago respectively, then they probably are not high quality python packages and need further improvements in general.

Regarding the updating of python packages for security patch releases -- I don't have any say in the policies enforced by by employer's digital products team and CISO so I have to do my best to make sure to not use various dependencies with high severity security issues. I could do this exclusively in my Slicer fork, but I aim to help pull in updates to the upstream to help the general Slicer open source community prepare for security risks.

**Comment by lassoan** [comment]
Blosc2 (will be nnunet dependency) needs Python-3.11 - https://github.com/lassoan/SlicerHDBrainExtraction/issues/3

**Comment by jamesobutler** [comment]
Reviewing blosc2’s code, the Python requirement is Python 3.11+ because they are following [SPEC0](https://scientific-python.org/specs/spec-0000/) per an inline comment. See the following commit where they dropped Python 3.10 support:
https://github.com/Blosc/python-blosc2/commit/f94349f6076a080952bef4915e701ef27341cb89

**Comment by jcfr** [comment]
Ditto. Once https://github.com/Slicer/Slicer/pull/8210 is finalized, I will be tackling the following:
* https://github.com/Slicer/Slicer/issues/7060
* https://github.com/Slicer/Slicer/issues/6388
* https://github.com/Slicer/Slicer/issues/7488

**Comment by lassoan** [comment]
[nnInteractive](https://github.com/MIC-DKFZ/nnInteractive) requires Python>=3.10. It looks very promising.

@jcfr what steps do you foresee to update Slicer's Python? What can we help with?

**Comment by jcfr** [comment]
This coming two weeks, I will be working on adding support to Python 3.12.

Credits go to preliminary research done by @sjh26 that will greatly help expedite the process 🙏

**Comment by lassoan** [comment]
Thanks for the quick update. Is there a branch that I can start testing? Is there anything I can help with?

**Comment by jamesgoldfarb** [comment]
Any updates on this? Willing to test....

5.9.0 today
import sys
print(sys.version)
3.9.10 (main, Apr 20 2025, 23:18:47) [MSC v.1942 64 bit (AMD64)]

**Comment by adamrankin** [comment]
Tagging to follow. Would also like to build/use 3.12 to load ROS2 python modules into Slicer

**Comment by jamesobutler** [comment]
Slicer has upgraded to Python 3.12 as of https://github.com/Slicer/Slicer/commit/458d435e9f32dee69df128279447de071be884f2.

**Comment by jcfr** [comment]
For future reference, the transition from Python `3.9.10` to `3.12.10` required the integration of over 50+ pull requests to python-cmake-build and multiple additional pull requests to `Slicer`.

### Slicer Updates

Pull requests integrated anticipating the transition to Python 3.12.10:

* https://github.com/Slicer/Slicer/pull/8454
* https://github.com/Slicer/Slicer/pull/8457
* https://github.com/Slicer/Slicer/pull/8459
* https://github.com/Slicer/Slicer/pull/8461
* https://github.com/Slicer/Slicer/pull/8467
* https://github.com/Slicer/Slicer/pull/8479

Follow-up pull requests addressing related regressions and improvements:

* https://github.com/Slicer/Slicer/pull/8489
* https://github.com/Slicer/Slicer/pull/8494
* https://github.com/Slicer/Slicer/pull/8548
* https://github.com/Slicer/Slicer/pull/8582
* https://github.com/Slicer/Slicer/pull/8632
* https://github.com/Slicer/Slicer/pull/8651
* https://github.com/Slicer/Slicer/pull/8654
* https://github.com/Slicer/Slicer/pull/8727


### python-cmake-buildsystem Updates

See https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/issues/350
